core/commands/object/patch.go

```diff
@@ -232,6 +232,21 @@ Use MFS and 'files' commands instead:
 			return err
 		}
 
+		// We do not allow producing blocks bigger than 1 MiB to avoid errors
+		// when transmitting them over BitSwap. The 1 MiB constant is an
+		// unenforced and undeclared rule of thumb hard-coded here.
+		modifiedNode, err := api.Dag().Get(req.Context, p.Cid())
+		if err != nil {
+			return err
+		}
+		modifiedNodeSize, err := modifiedNode.Size()
+		if err != nil {
+			return err
+		}
+		if modifiedNodeSize > 1024 * 1024 {
```

Comment:
My guess would be that returning an error wouldn't suffice. The node is already too big and this has to be reverted. But reverting it here is hard, because the reversal can itself cause another error. So I would think this must be rejected at the ObjectAPI level.

Reply:
@petar I see two places where we could potentially put this check deeper in the stack:

*  `(api *ObjectAPI) AddLink`: after the [`InsertNodeAtPath`](https://github.com/ipfs/go-ipfs/blob/51ca2192472286c5c96aa64e02f6325b1b5ddafa/core/coreapi/object.go#L216-L219) call (which has the actual 'insertion' logic) using a check similar to the one placed at the CLI level (fetch the modified node and check its size). We might also want to evaluate purging the modified copy from the repo.

* (deeper) in the `go-merkledag` repo executing the actual [insertion logic](https://github.com/ipfs/go-merkledag/blob/8794146d52c211b8af0456e734a248649e636be0/dagutils/utils.go#L122-L138): it has the advantage that already has a permanent/temporary store division where rejecting the modified node wouldn't impact the `go-ipfs` repo. The downside is that we would be injecting an arbitrary restriction on a library based on the limitation of another (bitswap) just because both are consumed by the same `go-ipfs` logic.

Either is doable so I'll leave to you the decision of where to execute.